capa/render/proto/__init__.py

```diff
@@ -5,440 +5,760 @@
+        behavior=pb.behavior,
+        method=pb.method,
+        id=pb.id,
+    )
+
+
+def maec_from_pb2(pb: capa_pb2.MaecMetadata) -> rd.MaecMetadata:
+    return rd.MaecMetadata(
+        analysis_conclusion=pb.analysis_conclusion or None,
+        analysis_conclusion_ov=pb.analysis_conclusion_ov or None,
+        malware_family=pb.malware_family or None,
+        malware_category=pb.malware_category or None,
+        malware_category_ov=pb.malware_category_ov or None,
+    )
+
+
+def rule_metadata_from_pb2(pb: capa_pb2.RuleMetadata) -> rd.RuleMetadata:
+    return rd.RuleMetadata(
+        name=pb.name,
+        namespace=pb.namespace or None,
+        authors=tuple(pb.authors),
+        scope=scope_from_pb2(pb.scope),
+        attack=tuple([attack_from_pb2(attack) for attack in pb.attack]),
+        mbc=tuple([mbc_from_pb2(mbc) for mbc in pb.mbc]),
+        references=tuple(pb.references),
+        examples=tuple(pb.examples),
+        description=pb.description,
+        lib=pb.lib,
+        is_subscope_rule=pb.is_subscope_rule,
+        maec=maec_from_pb2(pb.maec),
+    )
+
+
+def doc_from_pb2(doc: capa_pb2.ResultDocument) -> rd.ResultDocument:
+    rule_matches: Dict[str, rd.RuleMatches] = {}
+    for rule_name, matches in doc.rules.items():
+        m = rd.RuleMatches(
+            meta=rule_metadata_from_pb2(matches.meta),
+            source=matches.source,
+            matches=tuple([(addr_from_pb2(pair.address), match_from_pb2(pair.match)) for pair in matches.matches]),
+        )
+        rule_matches[rule_name] = m
+
+    return rd.ResultDocument(meta=metadata_from_pb2(doc.meta), rules=rule_matches)
+
+
+def main(argv=None):
+    if argv is None:
+        argv = sys.argv[1:]
+
+    parser = argparse.ArgumentParser(description="convert JSON result document to protobuf")
+    parser.add_argument("json_input", help="path to JSON result document to convert")
+    parser.add_argument("-j", "--json", action="store_true", help="emit JSON conversion of protobuf instead of text")
+    args = parser.parse_args(args=argv)
+
+    with open(args.json_input, "r", encoding="utf-8") as f:
+        fdata = f.read()
+
+    doc = rd.ResultDocument.parse_obj(json.loads(fdata))
+
+    proto_doc = doc_to_pb2(doc)
+
+    if args.json:
+        # TODO use ensure_ascii?
+        # including_default_value_fields -> so we get empty/unset fields
+        # see https://googleapis.dev/python/protobuf/latest/google/protobuf/json_format.html
+        json_obj = MessageToJson(
+            proto_doc, sort_keys=True, preserving_proto_field_name=True, including_default_value_fields=True
+        )
+        print(json_obj)
+    else:
+        print(proto_doc)
 
-    if definition["type"] == "string" and "enum" in definition:
-        emit_proto_enum(out, definition)
+    # TODO test?
+    # doc2 = rd.ResultDocument.parse_obj(json.loads(json_obj))
+    # doc2 = rd.ResultDocument.construct(json.loads(json_obj))
+    # assert doc == doc2
```

Comment:
also remove this

Reply:
[252f5ce](https://github.com/mandiant/capa/pull/1316/commits/252f5cebb7992455a0b1e2f2ce86285a3cd98bb7)